core/banking/fee_discount.go

```diff
@@ -0,0 +1,180 @@
+			return feeDiscountPerPartyAndAssetKeys[i].asset < feeDiscountPerPartyAndAssetKeys[j].asset
+		}
+
+		return feeDiscountPerPartyAndAssetKeys[i].party < feeDiscountPerPartyAndAssetKeys[j].party
+	})
+
+	for _, key := range feeDiscountPerPartyAndAssetKeys {
+		if _, ok := updatedKeys[key]; ok {
+			continue
+		}
+
+		// apply decay discount amount
+		e.feeDiscountPerPartyAndAsset[key] = e.decayFeeDiscountAmount(e.feeDiscountPerPartyAndAsset[key], assetQuantum)
+
+		updateDiscountEvents = append(updateDiscountEvents, events.NewTransferFeesDiscountUpdated(
+			ctx,
+			key.party,
+			assetID,
+			e.feeDiscountPerPartyAndAsset[key].Clone(),
+			e.currentEpoch,
+		))
+	}
+
+	e.broker.SendBatch(updateDiscountEvents)
+}
+
+func (e *Engine) ApplyFeeDiscount(ctx context.Context, asset string, party string, fee *num.Uint) (discountedFee *num.Uint, discount *num.Uint) {
+	discountedFee, discount = e.EstimateFeeDiscount(asset, party, fee)
+	if discount.IsZero() {
+		return discountedFee, discount
+	}
+
+	key := e.feeDiscountKey(asset, party)
+	defer e.broker.Send(
+		events.NewTransferFeesDiscountUpdated(ctx,
+			party, asset,
+			e.feeDiscountPerPartyAndAsset[key].Clone(),
+			e.currentEpoch,
+		),
+	)
+
+	e.feeDiscountPerPartyAndAsset[key].Sub(e.feeDiscountPerPartyAndAsset[key], discount)
+
+	return discountedFee, discount
+}
+
+func (e *Engine) EstimateFeeDiscount(asset string, party string, fee *num.Uint) (discountedFee *num.Uint, discount *num.Uint) {
+	if fee.IsZero() {
+		return fee, num.UintZero()
+	}
+
+	key := e.feeDiscountKey(asset, party)
+	accumulatedDiscount, ok := e.feeDiscountPerPartyAndAsset[key]
+	if !ok {
+		return fee, num.UintZero()
+	}
+
+	return calculateDiscount(accumulatedDiscount, fee)
+}
+
+func (e *Engine) AvailableFeeDiscount(asset string, party string) *num.Uint {
+	key := e.feeDiscountKey(asset, party)
+
+	if discount, ok := e.feeDiscountPerPartyAndAsset[key]; ok {
+		return discount.Clone()
+	}
+
+	return num.UintZero()
+}
+
+// decayFeeDiscountAmount update current discount with: discount x e.feeDiscountDecayFraction
+// or 0 if discount is less than e.feeDiscountMinimumTrackedAmount x quantum (where quantum is the asset quantum).
+func (e *Engine) decayFeeDiscountAmount(currentDiscount *num.Uint, assetQuantum num.Decimal) *num.Uint {
+	if currentDiscount.IsZero() {
+		return currentDiscount
+	}
+
+	decayedAmount := currentDiscount.ToDecimal().Mul(e.feeDiscountDecayFraction)
+
+	if decayedAmount.LessThan(e.feeDiscountMinimumTrackedAmount.Mul(assetQuantum)) {
+		return num.UintZero()
+	}
+
+	decayedAmountUint, _ := num.UintFromDecimal(decayedAmount)
+	return decayedAmountUint
+}
+
+func calculateDiscount(accumulatedDiscount, theoreticalFee *num.Uint) (discountedFee, discount *num.Uint) {
+	theoreticalFeeD := theoreticalFee.ToDecimal()
+	// min(accumulatedDiscount-theoreticalFee,0)
+	feeD := num.MinD(
+		accumulatedDiscount.ToDecimal().Sub(theoreticalFee.ToDecimal()),
```

Comment:
I will restructure it

Reply:
Ok. Hmm. The algorithm really works with negative numbers. It will not work with uints... Honestly I would probably just keep it. If there is ever issue with the computation (which I don't think it will since this probably wont' be called very often), we can change it. For now I would keep it.